src/software/ai/passing/receiver_position_generator.hpp

```diff
@@ -0,0 +1,285 @@
+     * receiving positions
+     */
+    explicit ReceiverPositionGenerator(
+        std::shared_ptr<const FieldPitchDivision<ZoneEnum>> pitch_division,
+        TbotsProto::PassingConfig passing_config);
+
+    /**
+     * Generates the best receiving positions for the friendly robots to go to
+     * @param world The world to generate the best receiving positions based on
+     * @param num_positions The number of receiving positions to generate
+     * @param existing_receiver_positions A set of existing receiver positions that will
+     * be avoided, if possible, when generating the new receiver positions.
+     * @param pass_origin_override An optional override for the position where the pass
+     * will be made from. If not provided, the ball position will be used.
+     * @return A vector of
+     *      min(num_positions, num_friendly_robots - existing_receiver_positions.size())
+     * positions that the receivers could use.
+     */
+    std::vector<Point> getBestReceivingPositions(
+        const World &world, unsigned int num_positions,
+        const std::vector<Point> &existing_receiver_positions = {},
+        const std::optional<Point> &pass_origin_override      = std::nullopt);
+
+
+   private:
+    /**
+     * Sample more points and update the best_receiving_positions map if better
+     * receiving positions in each zone were found
+     *
+     * @param world The world to sample receiving positions in
+     * @param zones_to_sample The subset of the zones to sample receiving positions in
+     * @param pass_origin The origin of the pass
+     * @param num_samples_per_zone The number of samples to take per zone
+     */
+    void updateBestReceiverPositions(const World &world, const Point &pass_origin,
+                                     const std::vector<ZoneEnum> &zones_to_sample,
+                                     unsigned int num_samples_per_zone);
+
+    // Pitch division
+    std::shared_ptr<const FieldPitchDivision<ZoneEnum>> pitch_division_;
+
+    // Passing configuration
+    TbotsProto::PassingConfig passing_config_;
+
+    std::map<ZoneEnum, PassWithRating> best_receiving_positions;
+    std::map<ZoneEnum, Point> prev_best_receiving_positions;
+
+    // TODO (NIMA): Added for debugging
+    std::vector<TbotsProto::DebugShapes::DebugShape> debug_shapes;
+
+    // A random number generator for use across the class
+    std::mt19937 random_num_gen_;
+
+    // The random seed to initialize the random number generator
+    static constexpr int RNG_SEED = 1010;
+};
+
+template <class ZoneEnum>
+ReceiverPositionGenerator<ZoneEnum>::ReceiverPositionGenerator(
+    std::shared_ptr<const FieldPitchDivision<ZoneEnum>> pitch_division,
+    TbotsProto::PassingConfig passing_config)
+    : pitch_division_(pitch_division),
+      passing_config_(passing_config),
+      random_num_gen_(RNG_SEED)
+{
+}
+
+template <class ZoneEnum>
+std::vector<Point> ReceiverPositionGenerator<ZoneEnum>::getBestReceivingPositions(
```

Comment:
the logic in this function is very hard to follow. consider breaking it down into smaller ones for readability.

Reply:
I tried refactoring it. Let me know if it still feels confusing